src/lib/swapper/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts

```diff
@@ -106,105 +110,102 @@ export async function getTradeQuote(
-          buyAsset,
-          intermediaryTransactionOutputs,
-          feeData: {
-            protocolFees,
-            networkFeeCryptoBaseUnit,
-          },
-          // might be better replaced by inputOutputRatio downstream
-          rate: estimateRate,
-          sellAmountIncludingProtocolFeesCryptoBaseUnit,
-          sellAsset,
-          sources: [
-            { name: `${selectedLifiRoute.steps[0].tool} (${SwapperName.LIFI})`, proportion: '1' },
-          ],
-        }
       }),
     )
+  }
 
-    const isSameChainSwap = sellAsset.chainId === buyAsset.chainId
+  // this corresponds to a "hop", so we could map the below code over selectedLifiRoute.steps to
+  // generate a multi-hop quote
+  const steps = await Promise.all(
+    selectedLifiRoute.steps.map(async lifiStep => {
+      // for the rate to be valid, both amounts must be converted to the same precision
+      const estimateRate = convertPrecision({
+        value: selectedLifiRoute.toAmountMin,
+        inputExponent: buyAsset.precision,
+        outputExponent: sellAsset.precision,
+      })
+        .dividedBy(bn(selectedLifiRoute.fromAmount))
+        .toString()
 
-    // The rate for the entire multi-hop swap
-    const netRate = convertPrecision({
-      value: selectedLifiRoute.toAmountMin,
-      inputExponent: buyAsset.precision,
-      outputExponent: sellAsset.precision,
-    })
-      .dividedBy(bn(selectedLifiRoute.fromAmount))
-      .toString()
-    // TODO(gomes): intermediary error-handling within this module function calls
-    return Ok({
-      minimumCryptoHuman: getMinimumCryptoHuman(
-        sellAssetPriceUsdPrecision,
-        isSameChainSwap,
-      ).toString(),
-      rate: netRate,
-      steps,
-      selectedLifiRoute,
-    })
-  } catch (e) {
-    // TODO(gomes): this is a temporary shim from the old error handling to monads, remove try/catch
-    if (e instanceof SwapError)
-      return Err(
+      const protocolFees = transformLifiStepFeeData({
+        chainId,
+        lifiStep,
+        assets,
+      })
+
+      const buyAmountCryptoBaseUnit = bnOrZero(selectedLifiRoute.toAmountMin)
+      const intermediaryTransactionOutputs = getIntermediaryTransactionOutputs(assets, lifiStep)
+
+      const networkFeeCryptoBaseUnit = await getNetworkFeeCryptoBaseUnit({
+        accountNumber,
+        chainId,
+        lifiStep,
+        supportsEIP1559,
+        wallet,
+      })
+
+      return {
+        allowanceContract: lifiStep.estimate.approvalAddress,
+        accountNumber,
+        buyAmountBeforeFeesCryptoBaseUnit: buyAmountCryptoBaseUnit.toString(),
+        buyAsset,
+        intermediaryTransactionOutputs,
+        feeData: {
+          protocolFees,
+          networkFeeCryptoBaseUnit,
+        },
+        // TODO(woodenfurniture): the rate should be top level not step level
+        // might be better replaced by inputOutputRatio downstream
+        rate: estimateRate,
```

Comment:
Or rather, this step-level key should be a step-level value, rather than the top-level rate. 

It's already the top-level rate living on the step object.

Reply:
agreed, i prefer that too. I'll update it

Reply:
agreed, updated the comment